possible address sanitizer bug

I decided to try address sanitizer on WSL Ubuntu today, but ran into 2 errors. These errors are only present when using the sanitizer, so my crate compiles normally. To give context there are 2 crates: `dylink_macro` and `dylink`. On `dylink_macro` there is a procedural macro called `dylink`, which is named the same as my `dylink` crate which I am re-exporting it to. I don't really know if this a bug or a problem on my end, but I thought I'd mention it just in case.

```rust
error[E0519]: the current crate is indistinguishable from one of its dependencies: it has the same crate-name `dylink_macro` and was compiled with the same `-C metadata` arguments. This will result in symbol conflicts between the two.
   --> src/lib.rs:104:9
    |
104 | pub use dylink_macro::dylink as dylink;
    |         ^^^^^^^^^^^^

error: cannot determine resolution for the attribute macro `crate::dylink`
  --> src/vulkan.rs:50:2
   |
50 |     crate::dylink(any(name = "libvulkan.so.1", name = "libvulkan.so"))
   |     ^^^^^^^^^^^^^
   |
   = note: import resolution is stuck, try simplifying macro imports
```

`cargo +nightly version`:
```
cargo 1.71.0-nightly (569b648b5 2023-05-05)
```

I included the full error output in a file below

[error.log](https://github.com/rust-lang/rust/files/11411842/error.log)

On `rustc 1.70.0-nightly (a266f1199 2023-03-22)` the error is:

```
error: undefined symbol: __asan_option_detect_stack_use_after_return
```

which isn't great, but at least suggests some issue related to asan.

On `rustc 1.70.0-nightly (1459b3128 2023-03-23)` the error is:

```
error[E0519]: the current crate is indistinguishable from one of its dependencies: it has the same crate-name `inline_python_macros` and was compiled with the same `-C metadata` arguments. This will result in symbol conflicts between the two.
```

Which is very confusing.

https://github.com/rust-lang/rust/pull/109213 was merged between those two versions and changes in which cases E0519 is emitted.

ping @oli-obk who wrote that PR.

--- a/compiler/rustc_metadata/src/creader.rs
+++ b/compiler/rustc_metadata/src/creader.rs
@@ -148,11 +148,15 @@ impl CStore {
         assert_eq!(self.metas.len(), self.stable_crate_ids.len());
         let num = CrateNum::new(self.stable_crate_ids.len());
         if let Some(&existing) = self.stable_crate_ids.get(&root.stable_crate_id()) {
-            let crate_name0 = root.name();
-            if let Some(crate_name1) = self.metas[existing].as_ref().map(|data| data.name()) {
+            // Check for (potential) conflicts with the local crate
+            if existing == LOCAL_CRATE {
+                Err(CrateError::SymbolConflictsCurrent(root.name()))
+            } else if let Some(crate_name1) = self.metas[existing].as_ref().map(|data| data.name())
+            {
+                let crate_name0 = root.name();
                 Err(CrateError::StableCrateIdCollision(crate_name0, crate_name1))
             } else {
-                Err(CrateError::SymbolConflictsCurrent(crate_name0))
+                Err(CrateError::NotFound(root.name()))
             }
         } else {
             self.metas.push(None);

--- a/compiler/rustc_metadata/src/locator.rs
+++ b/compiler/rustc_metadata/src/locator.rs
@@ -961,6 +961,7 @@ pub(crate) enum CrateError {
     DlSym(String),
     LocatorCombined(Box<CombinedLocatorError>),
     NonDylibPlugin(Symbol),
+    NotFound(Symbol),
 }
 
 enum MetadataError<'a> {
@@ -1130,6 +1131,18 @@ impl CrateError {
             }
             CrateError::NonDylibPlugin(crate_name) => {
                 sess.emit_err(errors::NoDylibPlugin { span, crate_name });
+            }
+            CrateError::NotFound(crate_name) => {
+                sess.emit_err(errors::CannotFindCrate {
+                    span,
+                    crate_name,
+                    add_info: String::new(),
+                    missing_core,
+                    current_crate: sess.opts.crate_name.clone().unwrap_or("<unknown>".to_string()),
+                    is_nightly_build: sess.is_nightly_build(),
+                    profiler_runtime: Symbol::intern(&sess.opts.unstable_opts.profiler_runtime),
+                    locator_triple: sess.opts.target_triple.clone(),
+                });
             }
         }
     }

--- a/tests/run-make/issue-83045/Makefile
+++ b/tests/run-make/issue-83045/Makefile
@@ -29,5 +29,5 @@ all:
 				  --crate-type=rlib \
 				  --edition=2018 \
 				  c.rs 2>&1 | tee $(TMPDIR)/output.txt || exit 0
-	$(CGREP) E0519 < $(TMPDIR)/output.txt
+	$(CGREP) E0463 < $(TMPDIR)/output.txt
 	$(CGREP) -v "internal compiler error" < $(TMPDIR)/output.txt

